lib/iris/coords.py

```diff
@@ -2775,6 +2779,366 @@ def xml_element(self, doc):
+            )
+
+        indices_min = indices.min()
+        if _lazy.is_lazy_data(indices_min):
+            indices_min = indices_min.compute()
+        if indices_min < self.start_index:
+            indices_error(
+                f"Lowest index: {indices_min} < start_index: {self.start_index} ."
+            )
+
+        indices_shape = indices.shape
+        if len(indices_shape) != 2:
+            indices_error(
+                f"Expected 2-dimensional shape, got: shape={indices_shape} ."
+            )
+
+        len_req_fail = False
+        element_lengths = self._lazy_element_lengths(indices, self.element_dim)
+        if self.cf_role_element in ("edge", "boundary"):
+            if (element_lengths != 2).any().compute():
+                len_req_fail = "len=2"
+        else:
+            if self.cf_role_element == "face":
+                min_size = 3
+            elif self.cf_role_element == "volume":
+                if self.cf_role_indexed_element == "edge":
+                    min_size = 6
+                else:
+                    min_size = 4
+            else:
+                raise NotImplementedError
+            if (element_lengths < min_size).any().compute():
+                len_req_fail = f"len>={min_size}"
+        if len_req_fail:
+            indices_error(
+                f"Not all elements meet requirement: {len_req_fail} - needed "
+                f"to describe '{self.cf_role}' ."
+            )
+
+    def switch_start_index(self):
+        # Don't use a normal property setter - would be inappropriate since
+        # other properties are also modified.
+        """
+        Change :attr:`start_index` to its alternative value and change the
+        :attr:`indices` values to match. (:attr:`start_index` can be either
+        ``0`` or ``1``).
+
+        """
+        new_index = 1 - self.start_index
+        index_change = new_index - self.start_index
+        self._metadata_manager.start_index = new_index
+        # Modify the indices values to correspond with the new start_index.
+        self._values += index_change
+
+    def switch_element_dim(self):
+        # Don't use a normal property setter - would be inappropriate since
+        # other properties are also modified.
+        """
+        Change :attr:`element_dim` to its alternative value and restructure
+        :attr:`indices` to match. (:attr:`element_dim` can be ``0`` or ``1``).
+
+        """
+        self._metadata_manager.element_dim = 1 - self.element_dim
+        # Modify the indices structure to match the new element_dim.
+        # We know that indices is 2-dimensional so can just swap the two axes.
+        new_indices = self._values.swapaxes(0, 1)
+        # Replace the data manager with one for the new shaped indices.
+        # This is safe to override since Connectivity is not attached to a cube
+        # dimension unlike other _DimensionalMetadata-using classes.
+        self._values_dm = DataManager(new_indices)
```

Comment:
Again, I don't see why we need this behaviour.

Plus, supporting this is really saying that `element_dim` isn't `read-only`... this is a setter in disguise.

Don't have it part of the contract. The alternative is that the users just creates a different `Connectivity` instance instead.

Reply:
We have agreed that:

* This is effectively a `transpose` method, and should be named as such
* To preserve read-only, it should return a new `Connectivity`, rather than being in-place.
   [This is in line with Numpy](https://numpy.org/doc/stable/reference/generated/numpy.transpose.html), and is also the way that `Cube.transpose()` _should_ work.

Reply:
97310e758a5b23a92c28910d885bfaa2aa2c79dd